src/transport/UTransport.cpp

```diff
@@ -41,40 +40,82 @@ v1::UStatus UTransport::send(const v1::UMessage& message) {
-                             std::optional<v1::UUri>&& source_filter) {
-	auto [sinkOk, reason1] = UriValidator::isValid(sink_filter);
-	if (!sinkOk) {
-		throw UriValidator::InvalidUUri(
-		    "sink_filter is not a valid URI |  " +
-		    std::string(UriValidator::message(*reason1)));
-	}
+UTransport::HandleOrStatus UTransport::registerListener(
+    ListenCallback&& listener, const v1::UUri& source_filter,
+    uint16_t sink_resource_filter) {
+	auto sink_filter = getEntityUri();
+	sink_filter.set_resource_id(sink_resource_filter);
+	return registerListener(std::move(listener), source_filter,
+	                        std::move(sink_filter));
+}
 
-	if (source_filter.has_value()) {
-		auto [srcOk, reason2] = UriValidator::isValid(source_filter.value());
-		if (!srcOk) {
+UTransport::HandleOrStatus UTransport::registerListener(
+    ListenCallback&& listener, const v1::UUri& source_filter,
+    std::optional<v1::UUri>&& sink_filter) {
+	if (!sink_filter) {
+		// Handle the special case of publish-like messages where only a source
+		// address is provided.
+		auto [source_ok, bad_source_reason] =
+		    UriValidator::isValidSubscription(source_filter);
+		if (!source_ok) {
+			throw UriValidator::InvalidUUri(
+			    "source_filter is not a valid URI |  " +
+			    std::string(UriValidator::message(*bad_source_reason)));
+		}
+	} else {
+		auto [source_ok, bad_source_reason] =
+		    UriValidator::isValidFilter(source_filter);
+		if (!source_ok) {
 			throw UriValidator::InvalidUUri(
 			    "source_filter is not a valid URI |  " +
-			    std::string(UriValidator::message(*reason2)));
+			    std::string(UriValidator::message(*bad_source_reason)));
+		}
+
+		auto [sink_ok, bad_sink_reason] =
+		    UriValidator::isValidFilter(*sink_filter);
+		if (!sink_ok) {
+			throw UriValidator::InvalidUUri(
+			    "sink_filter is not a valid URI |  " +
+			    std::string(UriValidator::message(*bad_sink_reason)));
 		}
 	}
 
 	auto [handle, callable] = CallbackConnection::establish(
 	    std::move(listener), [this](auto conn) { cleanupListener(conn); });
 
-	v1::UStatus status = registerListenerImpl(sink_filter, std::move(callable),
-	                                          std::move(source_filter));
+	v1::UStatus status = registerListenerImpl(
+	    std::move(callable), source_filter, std::move(sink_filter));
 
 	if (status.code() == v1::UCode::OK) {
 		return std::move(handle);
 	} else {
-		return uprotocol::utils::Unexpected(std::move(status));
+		return utils::Unexpected(std::move(status));
 	}
 }
 
-const v1::UUri& UTransport::getDefaultSource() const { return defaultSource_; }
+// NOTE: deprecated
+utils::Expected<UTransport::ListenHandle, v1::UStatus>
+UTransport::registerListener(const v1::UUri& sink_filter,
+                             ListenCallback&& listener,
+                             std::optional<v1::UUri>&& source_filter) {
+	if (!source_filter) {
+		// Workaround for deprecated API not fully reflecting the use cases
```

Comment:
I don't get this at all @gregmedd please IM me so we can walk through this please.

Reply:
@stevenhartley - I sent you a message, but we didn't manage to catch time to talk. We did go over this and discuss it amongst the developers and agreed that, while a bit ugly, this is probably the right way to handle the deprecated API.

The summary is that we are doing two things to fix the bug:

1. Deprecating the old API so it can be replaced with one that is better aligned with the spec
2. Modifying the behavior of the deprecated API so that it fixes the underlying bug in code that is currently using it.

We do not plan on keeping deprecated APIs forever, but in the short term this helps minimize the impact of changes we are making after alpha.2